Add a post on deploying vision models on Vertex AI (#467)
huggingface/blog: deploy-vertex-ai.md

      dtype: DT_FLOAT
      shape: (-1)
      name: StatefulPartitionedCall:0
  outputs['label'] tensor_info:
      dtype: DT_STRING
      shape: (-1)
      name: StatefulPartitionedCall:1
Method name is: tensorflow/serving/predict
```

The model will accept [<u>base64 encoded</u>](https://www.base64encode.org/) strings of images, perform
pre-processing, run inference, and finally perform the post-processing
steps. The strings are base64 encoded to prevent any
modifications during network transmission. Pre-processing includes
resizing the input image to 224x224 resolution, standardizing it to the
`[-1, 1]` range, and transposing it to the `channels_first` memory
layout. Postprocessing includes mapping the predicted logits to string
labels.

To perform a deployment on Vertex AI, you need to keep the model
artifacts in a [<u>Google Cloud Storage (GCS) bucket</u>](https://cloud.google.com/storage/docs/json_api/v1/buckets).
The accompanying Colab Notebook shows how to create a GCS bucket and
save the model artifacts into it.

# Deployment workflow with Vertex AI 

The figure below gives a pictorial workflow of deploying an already
trained TensorFlow model on Vertex AI.

> ![](./assets/97_vertex_ai/image7.png)

Reviewer: I really like the images in this blog post!

summarizes the request statistics:

![](./assets/97_vertex_ai/image5.png)

Among all the different statistics shown in the table, \`Average (ms)\`
refers to the average latency of the Endpoint. Locust fired off about
**17230 requests**, and the reported average latency is **646
Milliseconds**, which is impressive. In practice, you’d want to simulate
more real traffic by conducting the load test in a distributed manner.
Refer [<u>here</u>](https://cloud.google.com/architecture/load-testing-and-monitoring-aiplatform-models)
to learn more.

[<u>This directory</u>](https://github.com/sayakpaul/deploy-hf-tf-vision-models/tree/main/hf_vision_model_vertex_ai/locust)
has all the information needed to know how we conducted the load test.

# Pricing

You can use the [<u>GCP cost estimator</u>](https://cloud.google.com/products/calculator) to estimate the cost of usage, 
and the exact hourly pricing table can be found [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).
It is worth noting that you are only charged when the node is processing
the actual prediction requests, and you need to calculate the price with
and without GPUs.

> For the Vertex Prediction for a custom-trained model, we can choose

Reviewer: Pricing tends to be one of the things that can change, so I would suggest to keep this section a bit shorter/less precise and instead link more to external resources so this does not become outdated

granular manner. For example, each hour of each CPU core costs about
$0.03635495, and each hour of each GB or RAM costs about $0.0048783.

<div align="center">

| **Machine Type**            | **Hourly Pricing (USD)** |
|:-----------------------------:|:--------------------------:|
| n1-standard-8 (8vCPU, 30GB) | $ 0.4372                |

</div>

Also, when you attach accelerators to the compute node, you will get
extra charges by the type of the accelerator you want. We have used
`NVIDIA_TESLA_T4` for this blog post, but almost all modern
accelerators, including TPUs are supported. You can find further
information [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).

<div align="center">

| **Accelerator Type** | **Hourly Pricing (USD)** |
|:----------------------:|:--------------------------:|
| NVIDIA_TESLA_T4      | $ 0.4024                |
</div>

# Call for Action

The collection of TensorFlow vision models in 🤗 Transformers is growing. It now has support
state-of-the-art semantic segmentation with 
[SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer#transformers.TFSegformerForSemanticSegmentation).

> We encourage you to extend the deployment workflow you learned this post to semantic segmentation models like SegFormer.

Author: @deep-diver FYI.

Building a model registry that guarantees high availability and security
is nontrivial. Also, there is often a situation where you want to roll
back the current model to one of the past versions since we can not
control the inside of a black box machine learning model. Vertex AI
Model Registry allows us to achieve these without much difficulty.

The currently supported model types include `SavedModel` from
TensorFlow, scikit-learn, and XGBoost.

## Vertex AI Endpoint

From the user’s perspective, Vertex AI Endpoint simply provides an
endpoint to receive requests and send responses back. However, it has a
lot of things under the hood for machine learning operators to
configure. Here are some of the configurations that you can choose:

- Version of a model

- Specification of VM in terms of CPU, memory, and accelerators

- Min/Max number of compute nodes

- Traffic split percentage

- Model monitoring window length and its objectives

- Prediction requests sampling rate

# Performing the Deployment

The [`google-cloud-aiplatform`](https://pypi.org/project/google-cloud-aiplatform/)
Python SDK provides easy APIs to manage the lifecycle of a deployment on

> Vertex AI. It is divided into four steps - **1)** uploading a model,

Reviewer: +1 on making this a list

    to perform inference.

  - `accelerator_count` denotes the number of hardware accelerators to
    attach to each replica.

**Note** that providing an accelerator is not a requirement to deploy
models on Vertex AI.

Next, you deploy the endpoint using the above specifications:

```py
tf28_gpu_deployed_model = endpoint_service_client.deploy_model(
    endpoint=tf28_gpu_endpoint,
    deployed_model=tf28_gpu_deployed_model_dict,
    traffic_split={"0": 100},
).result()
```

Notice how you’re defining the traffic split for the model. If you had
multiple versions of the model, you could have defined a dictionary
where the keys would denote the model version and values would denote
the percentage of traffic the model is supposed to serve.

With a Model Registry and a dedicated
[<u>interface</u>](https://console.cloud.google.com/vertex-ai/endpoints)
to manage Endpoints, Vertex AI lets you easily control the important
aspects of the deployment.

It takes about 15 - 30 minutes for Vertex AI to scope the deployment.
Once it’s done, you should be able to see it on the
[<u>console</u>](https://console.cloud.google.com/vertex-ai/endpoints).

# Performing Predictions

> Assuming you succeeded in the above step, you can test the deployed

Reviewer 2: ```suggestion
If your deployment was successful, you can test the deployed
```

Endpoint by making a prediction request.

First, prepare a base64 encoded image string:

```py
import base64
import tensorflow as tf

image_path = tf.keras.utils.get_file(
    "image.jpg", "http://images.cocodataset.org/val2017/000000039769.jpg"
)
bytes = tf.io.read_file(image_path)
b64str = base64.b64encode(bytes.numpy()).decode("utf-8")
```

**4)** The following utility first prepares a list of instances (only

> one instance in this case) and then uses a prediction service client (of
> type [`PredictionServiceClient`](https://cloud.google.com/python/docs/reference/automl/latest/google.cloud.automl_v1beta1.services.prediction_service.PredictionServiceClient)).

Reviewer 2: ```suggestion
one instance in this case) and then uses a [`PredictionServiceClient`](https://cloud.google.com/python/docs/reference/automl/latest/google.cloud.automl_v1beta1.services.prediction_service.PredictionServiceClient)).
```

```
from google.protobuf import json_format
from google.protobuf.struct_pb2 import Value

def predict_image(image, endpoint, serving_input):
    # The format of each instance should conform to
    # the deployed model's prediction input schema.
    instances_list = [{serving_input: {"b64": image}}]
    instances = [json_format.ParseDict(s, Value()) for s in instances_list]

    print(
        prediction_service_client.predict(
            endpoint=endpoint,
            instances=instances,
        )
    )

predict_image(b64str, tf28_gpu_endpoint, serving_input)
```

For TensorFlow models deployed on Vertex AI, the request payload needs
to be formatted in a certain way. For models like ViT that deal with
binary data like images, they need to be base64 encoded. According to
the [<u>official guide</u>](https://cloud.google.com/vertex-ai/docs/predictions/online-predictions-custom-models#encoding-binary-data),
the request payload for each instance needs to be like so:

```py
{serving_input: {"b64": base64.b64encode(jpeg_data).decode()}}
```

The `predict_image()` utility prepares the request payload conforming
to this specification.

> If everything goes well with the deployment, if you call

Reviewer 2: ```suggestion
If everything goes well with the deployment, when you call
```

model, let’s now discuss some beneficial features provided by Vertex AI.
These help you get deeper insights into your deployment.

# Monitoring with Vertex AI

Vertex AI also lets you monitor your model without any configuration.
From the Endpoint console, you can get details about the performance of
the Endpoint and the utilization of the allocated resources.

![](./assets/97_vertex_ai/image8.png)

![](./assets/97_vertex_ai/image6.png)

As seen in the above chart, for a brief amount of time, the accelerator
duty cycle (utilization) was about 100% which is a sight for sore eyes.
For the rest of the time, there weren’t any requests to process hence
things were idle.

This type of monitoring helps you quickly flag the currently deployed
Endpoint and make adjustments as necessary. It’s also possible to
request monitoring of model explanations. Refer
[<u>here</u>](https://cloud.google.com/vertex-ai/docs/explainable-ai/overview)
to learn more.

# Local Load Testing

We conducted a local load test to better understand the limits of the
Endpoint with [<u>Locust</u>](https://locust.io/). The table below
summarizes the request statistics:

![](./assets/97_vertex_ai/image5.png)

> Among all the different statistics shown in the table \`Average (ms)\`

Reviewer 2: ```suggestion
Among all the different statistics shown in the table, \`Average (ms)\`
```

> refers to the average latency of the Endpoint. Locust fired about
> **17230 requests**, and their average latency is reported to be **646
> Milliseconds** which is impressive. In practice, you’d want to simulate

Reviewer 2: ```suggestion
refers to the average latency of the Endpoint. Locust fired off about
**17230 requests**, and the reported average latency is **646
Milliseconds**, which is impressive. In practice, you’d want to simulate
```

more real traffic by conducting the load test in a distributed manner.
Refer [<u>here</u>](https://cloud.google.com/architecture/load-testing-and-monitoring-aiplatform-models)
to learn more.

[<u>This directory</u>](https://github.com/sayakpaul/deploy-hf-tf-vision-models/tree/main/hf_vision_model_vertex_ai/locust)
has all the information needed to know how we conducted the load test.

# Pricing

We used the [<u>GCP cost estimator</u>](https://cloud.google.com/products/calculator) for this

> purpose, and the exact pricing table can be found [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).

Reviewer 2: I would clarify what `this purpose` refers to since it isn't very clear.

It is worth noting that you are only charged when the node is processing
the actual prediction requests, and you need to calculate the price with
and without GPUs.

For the Vertex Prediction for a custom-trained model, we can choose
[N1 machine types from `n1-standard-2` to `n1-highcpu-32`](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).
You used `n1-standard-8` for this post which is equipped with 8
vCPUs and 32GBs of RAM. The price of a node can be calculated in a more
granular manner. For example, each hour of each CPU core costs about
$0.03635495, and each hour of each GB or RAM costs about $0.0048783.

<div align="center">

| **Machine Type**            | **Hourly Pricing (USD)** |
|:-----------------------------:|:--------------------------:|
| n1-standard-8 (8vCPU, 30GB) | $ 0.4372                |

</div>

> Also, when you attach accelerators to the compute node, you will get
> extra charges by the type of the accelerator you want. We have used

Reviewer 2: ```suggestion
Also, when you attach accelerators to the compute node, you will be
charged extra by the type of accelerator you want. We used
```

`NVIDIA_TESLA_T4` for this blog post, but almost all modern
accelerators, including TPUs are supported. You can find further
information [<u>here</u>](https://cloud.google.com/vertex-ai/pricing#custom-trained_models).

<div align="center">

| **Accelerator Type** | **Hourly Pricing (USD)** |
|:----------------------:|:--------------------------:|
| NVIDIA_TESLA_T4      | $ 0.4024                |
</div>

# Call for Action

> The collection of TensorFlow vision models in 🤗 Transformers is growing. It now has support

Reviewer 2: ```suggestion
The collection of TensorFlow vision models in 🤗 Transformers is growing. It now supports
```

state-of-the-art semantic segmentation with 
[SegFormer](https://huggingface.co/docs/transformers/model_doc/segformer#transformers.TFSegformerForSemanticSegmentation).

> We encourage you to extend the deployment workflow you learned this post to semantic segmentation models like SegFormer.

Reviewer 2: ```suggestion
We encourage you to extend the deployment workflow you learned in this post to semantic segmentation models like SegFormer.
```

# Conclusion

> In this post, you learned how to deploy a Vision Transformer model using

Reviewer 2: ```suggestion
In this post, you learned how to deploy a Vision Transformer model with
```

the Vertex AI platform using the easy APIs it provides. You also learned

> how Vertex AI’s features benefit the model deployment process enabling
> you to focus on declarative configurations while taking away complex

Reviewer 2: ```suggestion
how Vertex AI’s features benefit the model deployment process by enabling
you to focus on declarative configurations and removing the complex
```
